File: October Challenge/Week_1/Java/Day_7/RotateList.java
// Problem : https://leetcode.com/problems/rotate-list/
// Solution:

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head==null || k==0) return head;
        ListNode ptr = head;
        ListNode copyHead = head;
        
        int len =1;
        while(ptr.next!=null){
            ptr=ptr.next;
            len++;
        }
        k=k%len;
        if(k==0) return head;
        k=len-k;
        ptr=head;
        while(k>1){
            ptr=ptr.next;
            k--;
        }
        head=ptr.next;
        ptr.next=null;
        
        ptr=head;
        while(ptr.next!=null){
            ptr=ptr.next;
        }
        ptr.next=copyHead;
        
        return head;
    }
}

File: October Challenge/Week_1/C++/Day_7/RotateList.cpp
class Solution {
	
	public:
	
	ListNode* rotateRight(ListNode* head, int k) {
    
    if(head==NULL)
    {
        return NULL;
    }
    
    int len=0;
    ListNode* temp=head;
    
    while(temp!=NULL)
    {
        len++;
        temp=temp->next;
    }
    
    k=k%len;
    
    if(k==0)
    {
        return head;
    }
    
    temp=head;
    
    for(int i=0;i<len-k-1;i++)
    {
        temp=temp->next;
    }
    
    ListNode* newhead=temp->next;
    temp->next=NULL;
    
    temp=newhead;
    
    while(temp->next)
    {
        temp=temp->next;
    }
    
    temp->next=head;
    
    return newhead;
    
}
};
